common/solver/batch_gmres_kernels.hpp.inc

```diff
@@ -0,0 +1,985 @@
+                // residual = abs(s(inner_iter + 1))
+                for (int c = threadIdx.x; c < res_norms_shared_entry.num_rhs;
+                     c += blockDim.x) {
+                    const uint32 conv = converged & (1 << c);
+
+                    if (conv) {
+                        continue;
+                    }
+
+                    res_norms_shared_entry.values[c] = abs(
+                        s_shared_entry
+                            .values[(inner_iter + 1) * s_shared_entry.stride +
+                                    c]);
+                }
+
+                __syncthreads();
+
+
+                const uint32 converged_prev = converged;
+
+                all_converged = stop.check_converged(
+                    outer_iter * (restart + 1) + inner_iter + 1,
+                    res_norms_shared_entry.values, {NULL, 0, 0, 0}, converged);
+
+
+                const uint32 converged_recent = converged_prev ^ converged;
+
+
+                // y = Hessenburg(0 : inner_iter,0 : inner_iter) \ s(0 :
+                // inner_iter) x = x + KrylovBasis(:, 0 : inner_iter ) * y
+                update_x(x_shared_entry, inner_iter,
+                         gko::batch::to_const(H_shared_entry),
+                         gko::batch::to_const(s_shared_entry),
+                         gko::batch::to_const(V_shared_entry),
+                         ~converged_recent);
+
+
+                logger.log_iteration(
+                    ibatch, outer_iter * (restart + 1) + inner_iter + 1,
+                    res_norms_shared_entry.values, converged);
+
+                if (all_converged) {
+                    __syncthreads();
+
+                    inner_loop_break_flag = true;
+
+                    break;
+                }
+            }
+
+            if (inner_loop_break_flag == true) {
+                break;
+            }
+
+            // y = Hessenburg(0:restart - 1,0:restart - 1) \ s(0:restart-1)
+            // x = x + KrylovBasis(:,0 : restart - 1) * y
+            update_x(x_shared_entry, restart - 1,
+                     gko::batch::to_const(H_shared_entry),
+                     gko::batch::to_const(s_shared_entry),
+                     gko::batch::to_const(V_shared_entry), converged);
+
+            // r = b
+            copy(b_global_entry, r_shared_entry, converged);
+            __syncthreads();
+            // r = r - A*x
+            advanced_spmv_kernel(static_cast<ValueType>(-1.0), A_global_entry,
+                                 gko::batch::to_const(x_shared_entry),
+                                 static_cast<ValueType>(1.0), r_shared_entry);
+            __syncthreads();
+
+
+            compute_norm2<ValueType>(gko::batch::to_const(r_shared_entry),
+                                     res_norms_temp_shared_entry);
+
+            __syncthreads();
+
+            copy(gko::batch::to_const(res_norms_temp_shared_entry),
+                 res_norms_shared_entry, converged);
```

Comment:
At some point we should discuss this. Is it so important for GMRES to stop the update on RHS that have converged? If we can just continue to update them, we can simplify and reduce the cost of the solver. Another way to address this would be for `compute_norm2`, `compute_dot` etc. to take a bit-mask and only update those RHS for whom the bit is set.

Reply:
Yeah, that is difficult to say. For correctness, I think it makes sense to stop updating the vectors which have converged. If for performance reasons not updating is better, then provided that the solutions are still correct, I guess it would be okay to continue to update them. But if possible, I think a bit-mask approach is definitely preferable.

Reply:
So now all batch dense functions have an overload which takes a bit-mask.

Reply:
Nice! Thanks